### PR TITLE
fix: private items are shown in completions for modules in fn body

### PR DESCRIPTION
**Pull request by dqkqd**
Close: #18287

### CODE REVIEW

File: crates/hir-def/src/visibility.rs [hunk 1]
```diff
@@ -138,6 +138,12 @@ impl Visibility {
...
         loop {
+            // `to_module` is not a block module and should not be moved upward,
+            // as doing so might expose private items.
+            if !to_module.is_block_module() {
+                break;
+            }
+
             match (to_module.block, def_map_block) {
                 // to_module is not a block, so there is no parent def map to use
                 (None, _) => (),
```
Comment (ShoyuVanilla) [inline crates/hir-def/src/visibility.rs:149]: Wasn't the case for the your above code `if !to_module.is_block_module() { .. }` being handled here?

Comment (Veykril) [reply crates/hir-def/src/visibility.rs:149]: Was wondering the same but actually not quite, `is_block_module` is defined as `self.block.is_some() && self.local_id == DefMap::ROOT`, so we are checking `self.block.is_none()` (the match arm part) or `self.local_id != ROOT` (thats the new thing).

So if I get this right, the problem is that if we are in a module within a block module we incorrectly move upwards out of the local module right now. (this is all so confusing ..,).

Can we encode that branch in the match as well (or clean up the branching in some other way)? This as proposed makes it difficult to read imo

Comment (dqkqd) [reply crates/hir-def/src/visibility.rs:143]: Thanks for pointing out.

Actually, the fix by checking whether `to_module` is a `block_module` was incorrect.

The problem was that we keep moving upwards even though we should stop as soon as we reach the same block as `from_module`'s `def_map`.

**Review by Veykril — APPROVED** [review]

**Comment by Veykril** [comment]
Thanks!
@bors r+

**Pull request merged**